[IMP+FIX] resource_booking: duration, location, involves me, hr_holidays_public

yajo opened the pull request:
This PR includes several improvements or fixes for the `resource_booking` module:

##  [IMP] resource_booking: use duration field instead of stop 

From now on, `resource.booking.type` duration represents the default duration that will be applied to new `resource.booking` of that type, instead of the hardcoded duration for every new booking.

This means that you can create a new booking in pending state and define a custom duration before sending the portal link to customer.

Thus, the available schedule slots in portal will be based on the type duration, whereas only those that have enough time left to complete the booking duration will be displayed.

![image](https://user-images.githubusercontent.com/973709/126973764-32d339c8-64c4-495f-8b14-f52747a0c69f.png)


##  [IMP] resource_booking: configurable auto-assigned combination and location

Thanks to this patch, you will be able to more predictably block the chosen resource combination, with the added checkbox.

Also, the location field is propagated to `resource.booking` records, and synced from there to `calendar.event` if needed. This way, you can alter the location before sending the portal invitation.

![image](https://user-images.githubusercontent.com/973709/126974005-e4e74578-5785-4ecf-843e-a4865657e3bb.png)


##  [FIX] resource_booking: "Involves me" filter 

The filter wasn't working fine due to the `auto_join=True` set in `meeting_id`. It was never displaying pending bookings.

Besides, it wasn't necessary to keep it as a field because it was displayed nowhere and the search could be done directly in the view. This way, there's less python code to maintain and we disable the possibility of having a negative domain, which enters doomed scenarios when x2many fields are involved (see odoo/odoo#43957).

##  [IMP] resource_booking: compatible with hr_holidays_public 

Adding `.with_context(exclude_public_holidays=True)` in needed places to let `hr_holidays_public` do its magic and never allow allocating a booking during a public holidays when that module is properly installed and configured.

![image](https://user-images.githubusercontent.com/973709/126974278-067d6738-03d9-4e6e-8ab4-7e7653332b6a.png)

## [IMP] resource_booking: optional name propagated to meeting

- Allow users to optionally define a name for the resource booking.
- Display that name in portal if set.
- Sync it with meeting name when creating it.


https://user-images.githubusercontent.com/973709/127126593-701f0fe7-1afb-41aa-a7c8-2f869a89e563.mp4



@Tecnativa TT30987

pedrobaeza commented on resource_booking/models/resource_booking.py:
```diff
@@ -231,31 +230,53 @@ def _compute_state(self):
+    @api.depends("meeting_id.duration", "type_id")
+    def _compute_duration(self):
+        """Compute duration for each booking."""
+        for one in self:
```
I don't like too much this notation of `one` and prefer `record`, that is more standard.

pedrobaeza reviewed (changes requested):
On a functional review, everything goes correct, but the booking confirmation shows the incorrect duration:

![Selección_014](https://user-images.githubusercontent.com/7165771/126672516-489a21a9-00a3-41d9-b345-4de340c7d530.png)

yajo commented:
All done. Please keep in mind that this will have to be merged with "nobump", as it includes the version bump needed to test migration scripts.

Commit:
[IMP] resource_booking: configurable auto-assigned combination and location

Thanks to this patch, you will be able to more predictably block the chosen resource combination, with the added checkbox.

Also, the location field is propagated to `resource.booking` records, and synced from there to `calendar.event` if needed. This way, you can alter the location before sending the portal invitation.

@Tecnativa TT30987

Commit:
[IMP] resource_booking: use duration field instead of stop

From now on, `resource.booking.type` duration represents the default duration that will be applied to new `resource.booking` of that type, instead of the hardcoded duration for every new booking.

This means that you can create a new booking in pending state and define a custom duration before sending the portal link to customer.

Thus, the available schedule slots in portal will be based on the type duration, whereas only those that have enough time left to complete the booking duration will be displayed.

@Tecnativa TT30987

Yajo commented:
Added new IMP to the PR:

> Thanks to this patch, you will be able to more predictably block the chosen resource combination, with the added checkbox.
>
> Also, the location field is propagated to `resource.booking` records, and synced from there to `calendar.event` if needed. This way, you can alter the location before sending the portal invitation.

Commit:
[FIX] resource_booking: "Involves me" filter

The filter wasn't working fine due to the `auto_join=True` set in `meeting_id`. It was never displaying pending bookings.

Besides, it wasn't necessary to keep it as a field because it was displayed nowhere and the search could be done directly in the view. This way, there's less python code to maintain and we disable the possibility of having a negative domain, which enters doomed scenarios when x2many fields are involved (see https://github.com/odoo/odoo/issues/43957).

@Tecnativa TT30987

Commit:
[IMP] resource_booking: compatible with hr_holidays_public

Adding `.with_context(exclude_public_holidays=True)` in needed places to let `hr_holidays_public` do its magic and never allow allocating a booking during a public holidays when that module is properly installed and configured.

@Tecnativa TT30987

yajo commented:
Updated description to reflect current state of the PR, which has 4 different feature sets. Ready to review/merge.

pedrobaeza commented on resource_booking/models/resource_booking.py:
```diff
@@ -100,14 +109,13 @@ class ResourceBooking(models.Model):
         copy=False,
         index=True,
-        inverse="_inverse_dates",
+        readonly=False,
```
Haven't you mixed these changes in an unrelated commit?

yajo replied:
AFAICS it's OK. I'm removing the inverse, but still allowing writes with readonly=False.

pedrobaeza replied:
I'm saying that the changes are on the second commit, while the topic belongs to the first commit.

yajo replied:
I double-checked it and it's OK. The need to change from `_inverse_dates` to `_sync_dates` happened in 1cd849d8c40f644aefb62391ce0c3dfe6dd78012.

pedrobaeza reviewed (commented):
This is still missing the `name` field on the resource.booking, needed for setting the meeting topic and to be transferred to the calendar.event.

yajo commented:
> This is still missing the `name` field on the resource.booking, needed for setting the meeting topic and to be transferred to the calendar.event.

Currently the name is computed based on the date, requester and type, and it's propagated to the calendar meeting.

That's on purpose because originally having to add a name would be uncomfortable and unnecessary, because basically every booking of the same type has the same purpose. Also, when placing a booking from e-commerce, we don't want the customer to have to write a name; that adds no value.

I can add the `name` field and make it optional. Then, if it's empty, I can compute the `display_name` to the same value `name` has now. Sounds good?

pedrobaeza commented:
I want to add a free text. For example: "Meeting about module X". And the title of the calendar.event should be that one, and in the portal that should be also. You tell me the technical specifications for that.

yajo commented:
Ok, but I mean that we need to make that optional. We can't require the name field to be filled, or we'll break the original purpose of the module.

pedrobaeza commented:
It can be optional indeed.

Commit:
[IMP] resource_booking: optional name propagated to meeting

- Allow users to optionally define a name for the resource booking.
- Display that name in portal if set.
- Sync it with meeting name when creating it.

@Tecnativa TT30987

yajo commented:
Custom optional name added (see last video on description).

pedrobaeza commented:
/ocabot merge nobump
